Xamarin.Forms.Core/Layout.cs

```diff
@@ -320,9 +320,7 @@ internal static void LayoutChildIntoBoundingRegion(View child, Rectangle region,
 
 		internal virtual void OnChildMeasureInvalidated(VisualElement child, InvalidationTrigger trigger)
 		{
-			ReadOnlyCollection<Element> children = LogicalChildrenInternal;
-			int count = children.Count;
-			for (var index = 0; index < count; index++)
+			for (var index = 0; index < LogicalChildrenInternal.Count; index++)
```

Comment:
by changing this, `LogicalChildrenInternal.Count` is evaluated at each iteration. please revert this part

Reply:
I've noticed that part, but there are many cases `LogicalChildrenInternal.Count` is being used in a for loop in the source code.

https://github.com/xamarin/Xamarin.Forms/blob/210df5f775b2aa682b2c823f1d854d8290e46e27/Xamarin.Forms.Core/StackLayout.cs#L62

https://github.com/xamarin/Xamarin.Forms/blob/49a937fc5c04d6f80d902217c4883d1370c72a4b/Xamarin.Forms.Core/Layout.cs#L260

https://github.com/xamarin/Xamarin.Forms/blob/28b343b2788fa073a3d3a87610275d8f3e756b61/Xamarin.Forms.Core/TemplatedView.cs#L21

https://github.com/xamarin/Xamarin.Forms/blob/5f2796e160f99bad31c1b06a2d1f328b5ed02ba3/Xamarin.Forms.Core/ContentPresenter.cs#L22

https://github.com/xamarin/Xamarin.Forms/blob/485b3450496c91dd7875bbe163a234ccb4c51df8/Xamarin.Forms.Core/Element.cs#L344

Do you still need this change?

Reply:
Since `LogicalChildrenInternal` is a `ReadOnlyCollection<Element>`, isn't the `Count` property a fixed `int`? Unless I'm missing something, I don't think there's any re-evaluation happening.

Reply:
I was looking up `ReadOnlyCollection<T>` last night. It doesn't seem all that intuitive, but based on what I read, this is a wrapper around an actual list/collection and doesn't allow changes to it. However, changes can still be made to the underlying type (by directly working with it) and they should be reflected back to `ReadOnlyCollection<T>`.

`InternalChildren` is an `ObservableCollection<Element>` and can change.

If the loop needs to read the count property only once, shouldn't all other loops be implemented the same way? Why were they left to read it continuously?

Reply:
> shouldn't all other loops be implemented the same way? Why were they left to read it continuously?

Don't discount the possibility that all those other loops are just wrong :)

I thought Stephane was referring to an issue like IEnumerable.Count(). But if the worry is the value of Count changing because the underlying collection changes, then it might make sense to cache it before the loop. Or perhaps he's worried about the speed of accessing the member during the loop.

Reply:
I actually prefer counts being cached for all those loops. Since a changing `Count` requires more thought process / planning, I **assume** when the loops were being written, people didn't think much of it, but I agree that Stephane should explain why he prefers it cached.